<launch>

    <!-- Arguments -->   
    <arg name="traj_name" default="horiz_search_quat" />   <!-- can only be a quaternion traj, not a flat output traj -->
    <arg name="do_land" default="true" />                <!-- land on the AprilTag or don't land at all (req. manual land from transmitter) -->    
    <arg name="do_fov" default="false" />                 <!-- apply FOV constraints to min-snap traj generation -->
    <arg name="use_percep_cost" default="false" />        <!-- use PAMPC, by applying perception cost to MPC -->    

    <arg name="land_spd_factor" default="0.45" />        <!-- Landing speed factor, smalling is longer landing time, 0 mean fixed landing speed idependent of current velocity -->
    <arg name="land_spd_base" default="0.43" />          <!-- avg land spd if land_spd_factor = 0 -->
    <arg name="land_height" default="0.1" />             <!-- stop motors when camera this high (m) above landing pad -->    
    <arg name="mpc_time_horizon" default="2.0" />        <!-- time horizon of acado MPC. From ../acado/PAMPC/quadrotor_pampc.cpp --> 
    <arg name="num_acado_iters" default="3" />           <!-- number real-time mpc iterations (first loop-through will always be max(num_acado_iters, 3)) -->
    <arg name="tag_smoothing_factor" default="0.01" />   <!-- (0,1] AprilTag smoothing factor. 1 means no smoothing -->    
    <arg name="half_horiz_fov" default="0.5925" />          <!-- half of down-facing horizontal (width) FOV (radians) (smaller than half_vert_fov) -->    
    <arg name="half_vert_fov" default="0.79" />           <!-- half of down-facing vertical (length) FOV (radians) -->    
    
    <arg name="simulation" default="true" />
    <arg name="cam_name" default="/pinhole_camera" />    <!-- for down-facing camera -->
    <arg name="img_topic" default="image_raw" />         <!-- for down-facing camera -->
    <arg name="debug_mpc" default="false" />
    <arg name="tran_BC_x" default="0.108" />             <!-- x translation from quad body to camera -->    
    <arg name="tran_BC_y" default="0.0" />               <!-- y translation from quad body to camera -->    
    <arg name="tran_BC_z" default="0.0" />               <!-- z translation from quad body to camera --> 
    
    <arg name="thrust_map_a" default="0.00013850414341400538" />   <!-- thrust map polynomial coefficients --> 
    <arg name="thrust_map_b" default="-0.005408755617324549" />     
    <arg name="thrust_map_c" default="0.11336157680888627" />       
    <arg name="thrust_map_d" default="-0.0022807568577082674" />    

    <!-- Controller node -->
    <node pkg="vioquad_land" name="mavros_cmd_node" type="mavros_cmd_node" output="screen" required="true">
        <param name="ctrl_mode" type="string" value="mpc" />
        <param name="enable_sim" type="bool" value="$(arg simulation)" />
    </node>

    <!-- MPC node -->
    <arg name="launch_prefix" value="xterm -e gdb --args" if="$(arg debug_mpc)" />
    <arg name="launch_prefix" value="" unless="$(arg debug_mpc)" />
    <node pkg="vioquad_land" name="mpc_node" type="mpc_node" output="screen" required="true" launch-prefix="$(arg launch_prefix)">
        <param name="search_traj_file" type="string" value="$(find vioquad_land)/extra/traj/traj_$(arg traj_name).csv" />
        <param name="land_spd_factor" value="$(arg land_spd_factor)" />
        <param name="land_spd_base" value="$(arg land_spd_base)" />
        <param name="land_height" value="$(arg land_height)" />
        <param name="do_land" value="$(arg do_land)" />
        <param name="do_fov" value="$(arg do_fov)" />
        <param name="mpc_time_horizon" value="$(arg mpc_time_horizon)" />
        <param name="num_acado_iters" value="$(arg num_acado_iters)" />

        <param name="tag_smoothing_factor" value="$(arg tag_smoothing_factor)" />
        <param name="use_percep_cost" value="$(arg use_percep_cost)" />
        <param name="half_horiz_fov" value="$(arg half_horiz_fov)" />
        <param name="half_vert_fov" value="$(arg half_vert_fov)" />
        
        <param name="enable_sim" type="bool" value="$(arg simulation)" />
        <param name="tran_BC_x" value="$(arg tran_BC_x)" />
        <param name="tran_BC_y" value="$(arg tran_BC_y)" />
        <param name="tran_BC_z" value="$(arg tran_BC_z)" />
        
        <param name="thrust_map_a" value="$(arg thrust_map_a)" />
        <param name="thrust_map_b" value="$(arg thrust_map_b)" />
        <param name="thrust_map_c" value="$(arg thrust_map_c)" />
        <param name="thrust_map_d" value="$(arg thrust_map_d)" />
    </node>

    <!-- Apriltag node -->
    <include file="$(find apriltag_ros)/launch/continuous_detection.launch">
        <arg name="camera_name" value="$(arg cam_name)"/>
        <arg name="image_topic" value="$(arg img_topic)"/>
    </include>

</launch>
